### PR TITLE
Added common Builder trait for EntityBuilder and LazyBuilder

### PR DESCRIPTION
**Pull request by Xt-8**
PR for #422.

Hopefully I did it right.

---
This change is [Reviewable](https://reviewable.io/reviews/slide-rs/specs/426)

### CODE REVIEW

**Comment by LinearZoetrope** [comment]
There should also probably be a serde-gated `MarkedBuilder` with `marked`.

**Comment by WaDelma** [comment]
This is breaking change indeed (as witnessed tests failing because they don't import the trait). I am not against doing one for this as it seems that it would be useful.

It probably needs the aforementioned `MarkedBuilder` which should be extension trait to the `Builder` trait.

**Comment by Xaeroxe** [comment]
Also we need to add `EntityResBuilder` to this, which was merged yesterday.

**Comment by GraemeWilde** [comment]
`EntityResBuilder` has an extra parameter in it's `with` (a storage), not sure what to there.

**Comment by Xaeroxe** [comment]
@GraemeWilde ah yeah.  Ok, just ignore it then.

**Comment by GraemeWilde** [comment]
I also updated the changelog and the specs version to 0.12.0 (I think thats what will need to happen?).

File: src/world/mod.rs [hunk 1]
```diff
@@ -76,15 +90,15 @@ pub struct EntityBuilder<'a> {
...
 }
 
-impl<'a> EntityBuilder<'a> {
+impl<'a> Builder for EntityBuilder<'a> {
     /// Appends a component and associates it with the entity.
     ///
     /// # Panics
     ///
     /// Panics if the component hasn't been `register()`ed in the
     /// `World`.
```
Comment (torkleyy) [inline src/world/mod.rs]: No need for the duplicated docs here.

File: src/saveload/marker.rs [hunk 2]
```diff
@@ -35,7 +61,7 @@ impl<'a> EntityBuilder<'a> {
     /// ## Panics
     ///
     /// Panics in case there's no allocator added to the `World`.
-    pub fn marked<M>(self) -> Self
+    fn marked<M>(self) -> Self
```
Comment (torkleyy) [inline src/saveload/marker.rs]: This still has duplicated docs

File: src/world/mod.rs [hunk 3]
```diff
@@ -97,7 +105,7 @@ impl<'a> EntityBuilder<'a> {
 
     /// Finishes the building and returns the entity.
```
Comment (torkleyy) [inline src/world/mod.rs]: "As opposed to `LazyBuilder`, the components are available immediately."

File: src/world/mod.rs [hunk 4]
```diff
@@ -34,6 +34,20 @@ impl<'a> Iterator for CreateIter<'a> {
     }
 }
 
+/// A common trait for EntityBuilder and LazyBuilder, allowing either to be used.
+pub trait Builder {
```
Comment (torkleyy) [inline src/world/mod.rs]: Might want to point out the guarantees here:

* entity is definitely alive
* components may or may not exist before a `maintain`

File: src/saveload/marker.rs (hunk 2 above)
Comment (GraemeWilde) [reply src/saveload/marker.rs]: Should I be removing the docs from the trait or the implementation? Do the docs from traits fall through to the implementations?

Comment (torkleyy) [reply src/saveload/marker.rs]: Docs on the implementation of a trait only make sense if there's something special about the implementation. So for the lazy stuff it does make sense to document the delay of the insertions.

**Comment by GraemeWilde** [comment]
Is it okay if I make all the doc mentions of `stuff` into links?

**Comment by torkleyy** [comment]
I used to do that but I don't anymore because the links don't work with
reexports.

**Comment by GraemeWilde** [comment]
Are you talking about when you `pub use` re-export something from a different crate?

**Comment by torkleyy** [comment]
@GraemeWilde Yes, but it doesn't even have to be in a different crate. If we reexport it in e.g. the prelude of Specs, the link will break.

File: src/world/mod.rs [hunk 5]
```diff
@@ -34,14 +34,16 @@ impl<'a> Iterator for CreateIter<'a> {
...
+/// Entity is definitely alive, but the components may or may not exist before a call to
+/// [`World::maintain`].
 pub trait Builder {
     /// Appends a component and associates it with the entity.
     ///
     /// # Panics
     ///
-    /// Panics if the component hasn't been `register()`ed in the
-    /// `World`.
+    /// Panics if the component hasn't been [`register()`](World::register)ed in the
```
Comment (torkleyy) [inline src/world/mod.rs]: Do these links really work? I never saw them like this.

Comment (GraemeWilde) [reply src/world/mod.rs]: Hmm. They work for nightly, but not stable apparently. The ``[`Thing`]`` ones without specifying the link later don't work without nightly either,. The re-export ones don't work (as you were saying) without `#[doc(no-inline)]`, but using that puts it in the re-exports section with no description...

If you would prefer, I'll just take out all the links.

**Pull request merged**